samza-core/src/main/java/org/apache/samza/util/EmbeddedTaggedRateLimiter.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.util;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.samza.config.Config;
+import org.apache.samza.task.TaskContext;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+
+/**
+ * An embedded rate limiter that supports tags
+ */
+public class EmbeddedTaggedRateLimiter implements RateLimiter {
+
+  private Map<String, Integer> tagToTargetRateMap;
+  private Map<String, com.google.common.util.concurrent.RateLimiter> tagToRateLimiterMap;
+
+  public EmbeddedTaggedRateLimiter(Map<String, Integer> tagToCreditsPerSecondMap) {
+    Preconditions.checkArgument(tagToCreditsPerSecondMap.size() > 0, "Map of tags can't be empty");
+    this.tagToTargetRateMap = tagToCreditsPerSecondMap;
+  }
+
+  @Override
+  public void acquire(Map<String, Integer> tagToCreditsMap) {
+    ensureTagsAreValid(tagToCreditsMap);
+    tagToCreditsMap.forEach((tag, numberOfCredits) -> tagToRateLimiterMap.get(tag).acquire(numberOfCredits));
+  }
+
+  @Override
+  public Map<String, Integer> acquire(Map<String, Integer> tagToCreditsMap, long timeout, TimeUnit unit) {
+    ensureTagsAreValid(tagToCreditsMap);
+
+    long timeoutInNanos = NANOSECONDS.convert(timeout, unit);
+
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    return tagToCreditsMap.entrySet().stream()
+        .map(e -> {
+            String tag = e.getKey();
```

Comment:
this basically serializes all the tryAquire() for credits. I wonder if we should issue all tryAquires in parallel so all tags get the whole timeout duration for credits. I think it makes more sense because there is no inherent relation among all the tags.

Reply:
I think there is some difference. Say we have compA trying to acquire 3 credits for tag1, and 2 credits for tag2. There are only 2 credits available for tag1 and 3 for tag2. So compA waits for tag1. While waiting, compB comes in and tries to get 3 credits for tag2, it gets the credit while compA is still waiting for tag1. Let's say refill has not happened for tag2, and compA just got the credits it needs for tag1, it will still have to wait for credits for tag2 even though it is ahead of line w.r.t. compB. If we do parallel wait, then compA will only have to wait for tag1, and can get teh credits before compB does. 

Does this make sense?

Reply:
Well, I understand there is different dynamics. However, do you see a difference in the aggregated effect? Let's chat face to face then.

Reply:
I am not sure about the aggregated effect. I am fine with the current approach, just throwing out some ideas in case it makes sense.